File: redhawk/codegen/jinja/python/component/base/templates/properties.py
#{% macro simple(prop) %}
#{%   filter codealign %}
${prop.pyname} = simple_property(id_="${prop.identifier}",
#%    if prop.name
                                 name="${prop.name}",
#%    endif
                                 type_="${prop.type}",
#%    if prop.pyvalue is defined
                                 defvalue=${prop.pyvalue},
#%    endif
#%    if prop.isComplex
                                 complex=True,
#%    endif
                                 mode="${prop.mode}",
                                 action="${prop.action}",
                                 kinds=${prop.kinds|quote|tuple}
#%-    if prop.hasDescription
,
                                 description="""${prop.description}"""
#%    endif
                                 )
#{%   endfilter %}
#{% endmacro %}

#{% macro simplesequence(prop) %}
#{%   filter codealign %}
${prop.pyname} = simpleseq_property(id_="${prop.identifier}",
#%    if prop.name
                                    name="${prop.name}",
#%    endif
                                    type_="${prop.type}",
#%    if prop.pyvalue
                                    defvalue=[
#{%   for val in prop.pyvalue %}
                                              ${val},
#{%   endfor %}
                                             ],
#%    else
                                    defvalue=None,
#%    endif
#%    if prop.units
                                    units="${prop.units}",
#%    endif
#%    if prop.isComplex
                                    complex=True,
#%    endif
                                    mode="${prop.mode}",
                                    action="${prop.action}",
                                    kinds=${prop.kinds|quote|tuple}
#%-    if prop.hasDescription
,
                                    description="""${prop.description}"""
#%    endif
                                    )
#{%   endfilter %}
#{% endmacro %}

#{% macro struct(prop) %}
#{%   filter codealign %}
${prop.pyname} = struct_property(id_="${prop.identifier}",
#%    if prop.name
                                 name="${prop.name}",
#%    endif
                                 structdef=${prop.pyclass},
                                 configurationkind=${prop.kinds|quote|tuple},
                                 mode="${prop.mode}"
#%-    if prop.hasDescription
,
                                 description="""${prop.description}"""
#%    endif
                                 )
#{%   endfilter %}
#{% endmacro %}

#{% macro structsequence(prop) %}
#{%   filter codealign %}
${prop.pyname} = structseq_property(id_="${prop.identifier}",
#%    if prop.name
                                    name="${prop.name}",
#%    endif
                                    structdef=${prop.structdef.pyclass},
                                    defvalue=[${prop.pyvalues|join(',')}],
                                    configurationkind=${prop.kinds|quote|tuple},
                                    mode="${prop.mode}"
#%-    if prop.hasDescription
,
                                 description="""${prop.description}"""
#%    endif
                                    )
#{%   endfilter %}
#{% endmacro %}

#{% macro initializer(fields) %}
#{%   filter trim|lines|join(', ') %}
#{%   for field in fields %}
${field.pyname}=${field.pyvalue|default(python.defaultValue(field.type))}
#{%   endfor %}
#{%   endfilter %}
#{% endmacro %}

#{% macro members(fields) %}
#{%   filter trim|lines|join(',') %}
#{%   for field in fields %}
("${field.pyname}",self.${field.pyname})
#{%   endfor %}
#{%   endfilter %}
#{% endmacro %}

#{% macro structdef(struct, initialize=true) %}
class ${struct.pyclass}(object):
#{%   for field in struct.fields %}
#{%   filter codealign %}
    ${field.pyname} = simple_property(id_="${field.identifier}",
#{%     if field.name %}
                                      name="${field.name}",
#{%     endif %}
                                      type_="${field.type}",
#{%     if field.pyvalue %}
                                      defvalue=${field.pyvalue},
#{%     endif %}
#{%     if field.isComplex %}
                                      complex=True,
#{%     endif %}
                                      )
#{%   endfilter %}
#{%   endfor %}

#{%   if initialize: %}
    def __init__(self, **kw):
        """Construct an initialized instance of this struct definition"""
        for attrname, classattr in type(self).__dict__.items():
            if type(classattr) == simple_property:
                classattr.initialize(self)
        for k,v in kw.items():
            setattr(self,k,v)
#{%   else %}
    def __init__(self, ${initializer(struct.fields)}):
#{%     for field in struct.fields %}
        self.${field.pyname} = ${field.pyname}
#{%     endfor %}
#{%   endif %}

    def __str__(self):
        """Return a string representation of this structure"""
        d = {}
#{%   for field in struct.fields %}
        d["${field.pyname}"] = self.${field.pyname}
#{%   endfor %}
        return str(d)

    def getId(self):
        return "${struct.identifier}"

    def isStruct(self):
        return True

    def getMembers(self):
        return [${members(struct.fields)}]
#{% endmacro %}

File: redhawk/codegen/jinja/java/component/base/templates/properties.java
/*{% macro header(prop) %}*/
/**
 * The property ${prop.identifier}
 * ${prop.description|default("If the meaning of this property isn't clear, a description should be added.", true)}
 *
 * @generated
 */
/*{%- endmacro %}*/

/*{% macro structdef(prop) %}*/
/**
 * The structure for property ${prop.identifier}
 * 
 * @generated
 */
public static class ${prop.javatype} extends StructDef {
/*{% for field in prop.fields %}*/
    ${simple(field)|indent(4)}
/*{% endfor %}*/

    /**
     * @generated
     */
    public ${prop.javatype}(
/*{%- filter trim|lines|join(', ') %}*/
/*{%   for field in prop.fields %}*/
${field.javatype} ${field.javaname}
/*{%   endfor %}*/
/*{% endfilter -%}*/
) {
        this();
/*{% for field in prop.fields %}*/
        this.${field.javaname}.setValue(${field.javaname});
/*{% endfor %}*/
    }

    /**
     * @generated
     */
    public ${prop.javatype}() {
/*{% for field in prop.fields %}*/
        addElement(this.${field.javaname});
/*{% endfor %}*/
    }
};
/*{%- endmacro %}*/

/*{% macro simple(prop) %}*/
${header(prop)}
public final ${prop.javaclass}Property ${prop.javaname} =
    new ${prop.javaclass}Property(
        "${prop.identifier}", //id
        ${java.stringLiteral(prop.name) if prop.name else java.NULL}, //name
        ${prop.javavalue}, //default value
        Mode.${prop.mode|upper}, //mode
        Action.${prop.action|upper}, //action
        new Kind[] {${prop.javakinds|join(',')}} //kind
        );
/*{% endmacro %}*/

/*{% macro simplesequence(prop) %}*/
${header(prop)}
public final ${prop.javaclass}SequenceProperty ${prop.javaname} =
    new ${prop.javaclass}SequenceProperty(
        "${prop.identifier}", //id
        ${java.stringLiteral(prop.name) if prop.name else java.NULL}, //name
        ${prop.javaclass}SequenceProperty.asList(${prop.javavalues|join(',')}), //default value
        Mode.${prop.mode|upper}, //mode
        Action.${prop.action|upper}, //action
        new Kind[] {${prop.javakinds|join(',')}} //kind
        );
/*{% endmacro %}*/

/*{% macro struct(prop) %}*/
${structdef(prop)}

${header(prop)}
public final StructProperty<${prop.javatype}> ${prop.javaname} =
    new StructProperty<${prop.javatype}>(
        "${prop.identifier}", //id
        ${java.stringLiteral(prop.name) if prop.name else java.NULL}, //name
        ${prop.javatype}.class, //type
        new ${prop.javatype}(), //default value
        Mode.${prop.mode|upper}, //mode
        new Kind[] {${prop.javakinds|join(',')}} //kind
        );
/*{% endmacro %}*/

/*{% macro structsequence(prop) %}*/
${structdef(prop.structdef)}

${header(prop)}
public final StructSequenceProperty<${prop.structdef.javatype}> ${prop.javaname} =
    new StructSequenceProperty<${prop.structdef.javatype}> (
        "${prop.identifier}", //id
        ${java.stringLiteral(prop.name) if prop.name else java.NULL}, //name
        ${prop.structdef.javatype}.class, //type
/*{% if prop.javavalues %}*/
        StructSequenceProperty.asList(
/*{%   filter trim|lines|join(',\n')|indent(12, true) %}*/
/*{%     for values in prop.javavalues %}*/
            new ${prop.structdef.javatype}(${values|join(', ')})
/*{%     endfor %}*/
/*{%   endfilter %}*/

            ), //defaultValue
/*{% else %}*/
        StructSequenceProperty.<${prop.structdef.javatype}>asList(), //defaultValue
/*{% endif %}*/
        Mode.${prop.mode|upper}, //mode
        new Kind[] { ${prop.javakinds|join(',')} } //kind
    );
/*{% endmacro %}*/

/*{% macro structsequence_init(prop) %}*/
/*{% endmacro %}*/
